AzureIoTHub.agent.lib.nut

```diff
@@ -629,513 +632,819 @@ class AzureIoTHub {
+            } else if (topic.find(_topics.dMethodNotif) != null) {
+                _handleDirMethodMsg(message, topic);
+            }
         }
 
-        function _sendEvent(message, done) {
-            // create transfer and send
-            _transfers.event <- _senders.event.createtransfer(message.createAMQPMessage());
-            _log("send event transfer created")
-            _transfers.event.sendasync(function() {
-                _log("in send event transfer callback");
-                if (done) done(null);
-            }.bindenv(this));
+        function _handleCloudToDevMsg(message, topic) {
+            local props = {};
+            local splittedTopic = split(topic, "/");
+            if (splittedTopic[splittedTopic.len() - 1] != "devicebound") {
+                // We have properties sent with message
+                props = http.urldecode(splittedTopic[splittedTopic.len() - 1]);
+            }
+            try {
+                _onMessageCb(AzureIoTHub.Message(message, props));
+            } catch (e) {
+                _error("Exception at onReceive user callback: " + e);
+            }
         }
 
-        function _handleDeliveries(deliveries, cb) {
-            while (deliveries.len()) {
-                local item = AzureIoTHub.Delivery(deliveries.remove(0));
-                cb(null, item);
+        function _handleDesPropsMsg(message, topic) {
+            local version = null;
+            local parsedMsg = null;
+            try {
+                version = split(topic, "=")[1];
+                parsedMsg = http.jsondecode(message);
+            } catch (e) {
+                _error("Exception at parsing the message: " + e);
+                _logMsg(message, topic);
+                return;
+            }
+            try {
+                _onTwinReqCb(version, parsedMsg);
+            } catch (e) {
+                _error("Exception at onRequest user callback: " + e);
             }
         }
 
-        function _log(logMsg) {
-            if (_debug) server.log(logMsg);
+        function _handleTwinReqResMsg(message, topic) {
+            local status = null;
+            local reqId = null;
+            local parsedMsg = null;
+            try {
+                local splitted = split(topic, "/");
+                status = splitted[3].tointeger();
+                reqId = http.urldecode(splitted[4])["?$rid"];
+                if (_statusIsOk(status) && message != "") {
+                    parsedMsg = http.jsondecode(message);
+                }
+            } catch (e) {
+                _error("Exception at parsing the message: " + e);
+                _logMsg(message, topic);
+                return;
+            }
+            // Twin's properties received after UPDATE request
+            if (reqId in _twinUpdCbMap) {
+                local cb = _twinUpdCbMap[reqId][0];
+                local timer = _twinUpdCbMap[reqId][1];
+                delete _twinUpdCbMap[reqId];
```

Comment:
delete returns the deleted entry. you could cache it locally on delete and then reuse instead of accessing it three time in a row

Reply:
Done